### DISCUSSION TITLE
Problem with Array with multiple nested arrays

### DISCUSSION THREAD

**Discussion by julio3266** [react-hook-form/react-hook-form#7737] [Show and tell] [answered]
Hi, I was wondering if anyone could help me with a problem. I'm building a form that receives data from an array that has the following structure:

```
const arrayValues = [
   [
         ['ADULT'], ['CHILD'] // here are the values ​​for each input, it only reads these values ​​from above
   ], 
   [
      ['ADULT'], ['CHILD']
   ]
]
```

however when giving console in the useForm control it reads only the first array, the form is being rendered correctly, the problem is in the value of each input, because it is only reading the data from an array.

**Comment by Moshyfawn** [comment]
Can you provide [a codesandbox](https://codesandbox.io/s/react-hook-form-v7-ts-template-4tn8i) for your use-case?

**Reply by julio3266** [reply]
of course... just five minutes, please

**Reply by julio3266** [reply]
@Moshyfawn https://codesandbox.io/s/react-hook-form-js-forked-rby4h?file=/src/App.js

**Comment by Moshyfawn** [comment, marked as answer]
So, there're a couple of issue with what you're trying to do.
- The main one is that you're using RHF V7 _(7.22.5)_
```json
"dependencies": {
  "react-hook-form": "7.22.5"
},
```
but the way you're using the `<Controller />` is from V6
```js
render={({ value, onChange }) => (
  <input value={value} onChange={onChange} />
)}
```
Since V7, the API for `<Controller />` has changed and now requires you to access your field value and handlers from the `field` property returned from `render`
```js
render={({ field }) => (
  <input value={field.value} onChange={field.onChange} />
)}
```
Please, refer to [the RHF V7 docs](https://react-hook-form.com/api/usecontroller/controller) to learn more about the current API
- Second point is RHF doesn't support flat arrays. What that means is you can't have this structure:
```js
const DATA = [
  [
    {
      type: "ADULT"
    }
  ],
  [
    {
      prop: "ADULT"
    }
  ]
]
```
You gotta extend it to something like this:
```js
const defaultValues = {
  rooms: [
    {
      number: 101,
      guests: [
        { type: "ADULT", firstName: "", lastName: "" },
        { type: "ADULT", firstName: "", lastName: "" }
      ]
    },
    {
      number: 102,
      guests: [
        { type: "ADULT", firstName: "", lastName: "" },
        { type: "ADULT", firstName: "", lastName: "" }
      ]
    }
  ]
};
```
Note the `rooms` and `guest` properties.
- The third point is more of a suggestion. I encourage you to explore and use[ the `useFieldArray` form hook](https://react-hook-form.com/api/usefieldarray) instead of manually trying to manage your nested values. It will help you `add` / `remove` etc. your fields if you need such functionality in the future _(I assume that from the data you're working with)_.

I've created [this CSB](https://codesandbox.io/s/react-hook-form-js-forked-ugl5s?file=/src/App.js:173-547) with the concepts I just talked about above. Please, take a look and don't hesitate to ask for clarifications

**Reply by julio3266** [reply]
In the project I'm working on we use v6

**Reply by Moshyfawn** [reply]
You've provided CSB with the `7.22.5` version installed ;) There's a difference in the API between these versions
All the same applies accept for the `<Controller />` API part, tho

**Reply by julio3266** [reply]
ah... this is enlightening, I'm going to test with this array structure, it's that in the project we have the room listing and the forms listing and this comes from an api, as it will render

**Reply by julio3266** [reply]
Fiz o que você passou e renderizou corretamente, porém quando manipulei os dados vindos da api, o useFieldArray permanence como array(0)
 first useFieldArray: 
<img width="729" alt="Captura de Tela 2022-02-04 às 10 32 21" src="https://user-images.githubusercontent.com/52897217/152537943-85de4482-91be-4083-9b6a-d5913dbbc360.png">

normalized data:
<img width="547" alt="Captura de Tela 2022-02-04 às 10 34 07" src="https://user-images.githubusercontent.com/52897217/152537987-4c8a7774-3587-4e03-8f9f-37138c61dba2.png">

fieldArray:
<img width="872" alt="Captura de Tela 2022-02-04 às 10 37 32" src="https://user-images.githubusercontent.com/52897217/152538205-136e90cc-81a7-4c3b-b3fc-9d046b03254e.png">

**Reply by Moshyfawn** [reply]
I don't speak Portuguese, but from what I could translate, you're saying that your data comes from BE, so it's async loaded after the form has been initialized, correct?
If I understand you correctly, you want to call [the `reset` form method](https://react-hook-form.com/api/useform/reset) with your normalized `defaultValues` after the data has come from BE to set the new `defaultValues`, because the form has already been initialized without it _(or with the first `defaultValues` you've provided)_.

```js
const { reset } = useForm()

const { normalizedAsyncData } = useAsyncData()

useEffect(() => {
  if (normalizedAsyncData) {
    reset(normalizedAsyncData)
  }
}, [reset, normalizedAsyncData])

```

**Reply by julio3266** [reply]
I'm so sorry, I forgot of typing in english

**Reply by Moshyfawn** [reply]
No worries, it's 2022 and tech allows us to communicate either way ;)

**Reply by julio3266** [reply]
hahaha exactly. I'm gonna test  this sugestion

**Reply by julio3266** [reply]
Thx, my problem is closed <3

**Reply by Moshyfawn** [reply]
Nice to hear, cheers!

**Reply by bluebill1049** [reply]
thanks, @Moshyfawn for the detailed reply and help!

**Reply by julio3266** [reply]
hey @Moshyfawn  one last question, can i insert the validations using the yup resolver in useFieldArray? because the form is in the second component

**Reply by bluebill1049** [reply]
you can use either in build or schema validation.

**Reply by Moshyfawn** [reply]
There's no mechanism to validate the `useFieldArray` directly ATM.
You can try validating your field array by creating a side-effect and use `setError` on certain conditions.
```js
useEffect(() => {
  if (fields.length > 10) {
    setError()
  }
}, [fields, setError])
```
Be aware that there's a limitation in RHF that if you're using TS, it will complain if set error to a non existent field. You can read about it more in #7661

You can switch to the form level validation with [@hookform/resolvers](https://github.com/react-hook-form/resolvers).

Take note of [this rule](https://react-hook-form.com/api/useform#validationResolver)
> A resolver cannot be used with the built-in validators (e.g.: required, min, etc.)

**Reply by julio3266** [reply]
@Moshyfawn I understand that there are limitations, but since I have a separate form on another useForm screen, and it generates multiple arrays, would it be possible to generate an error for each different entity?

**Reply by Moshyfawn** [reply]
@julio3266 I don't think I exactly understand what you're asking. Can you elaborate or give an example?
Do you mean to append an error to the "root" array object? You can still do that with your validation schema or [the `setError` form method](https://react-hook-form.com/api/useform/seterror). See [this CSB](https://codesandbox.io/s/awesome-zeh-g8yeq?file=/src/App.tsx)

**Reply by julio3266** [reply]
@Moshyfawn I'm using that CSB you sent, with basically the same structure, but I need to validate each field/form

**Reply by julio3266** [reply]
in this CSB 
https://codesandbox.io/s/react-hook-form-js-forked-ugl5s?file=/src/App.js:173-547

**Reply by Moshyfawn** [reply]
Again, I'm not sure I understand what you're trying to do.
If you wanna validate the fields using your validation schema, you can just add your validation conditions to your fields
```js
const itemSchema = z.object({
  name: z.string({ required_error: "Please, enter an item name" })
});
```
Each array is being validated against the validation schema as well
```js
const formSchema = z.object({
  list: z.array(itemSchema).min(10, "Please, add at least 10 list items"),
  array: z.array(itemSchema).min(5, "Please, add at least 5 array items")
});
```
If you don't use a validation schema, you can use [the `setError` method](https://react-hook-form.com/api/useform/seterror) with the `useEffect` to validate on value `onChange` and for certain conditions `setError` with the name of your field array to validate like your array length or smth. Again, it has the limitations described above. The array fields you can validate same as every other field: buy providing `rules` to the `Controller` or passing an object of your validation conditions as the second argument to `register`

**Reply by Moshyfawn** [reply]
Is this what you're looking for, @julio3266? I've updated [my CSB](https://codesandbox.io/s/react-hook-form-js-forked-ugl5s?file=/src/App.js) with some validation rules and error messages.

**Reply by julio3266** [reply]
@Moshyfawn just to understand... i need use the validation in Controller or i can use the resolver  as yup schema in this case...

**Reply by JamesWong1999** [reply]
I think validation should be either with build-in or schema validation, they can't be mixed.

**Reply by Moshyfawn** [reply]
Yep, @JamesWong1999, you're right!
@julio3266, you wanna decided to go with field level validations (`rules` on `Controller` and validation on `register`) or to use the resolver and validation schema. I'd say, resolvers are quite a bit more flexible. Seeing your use-case, it's easier to manage the more complex forms with a validation schema.

**Reply by julio3266** [reply]
@Moshyfawn, that it! The problem is, i can't use the useForm in the child component(form). How i can insert a resolver, in this case?

**Comment by RhydamKapoor** [comment]
{
      name: "",
      description: "",
      data: [
        {
          id: "",
          heading: "",
          line: [{ id: "", value: "" }],
          code: [{ id: "", value: "" }],
        },
      ],
    },
    
    How will you access line and code? How will you append or remove in this (line and code) array in the data array loop?
